mlflow/sklearn/__init__.py

```diff
@@ -455,6 +461,270 @@ def load_model(model_uri):
+       the one which generate the prediction result) and run_id.
+    (2) _log_post_training_metrics_enabled flag, in the following method scope:
+       `model.fit`, `eval_and_log_metrics`, `model.score`,
+       in order to avoid nested/duplicated autologging metric, when run into these scopes,
+       we need temporarily disable the metric autologging.
+    (3) eval_dataset_info_map, it is a double level map:
+       `eval_dataset_info_map[run_id][eval_dataset_var_name]` will get a list, each
+       element in the list is an id of "eval_dataset" instance.
+       This data structure is used for:
+        * generating unique dataset name key when autologging metric. For each eval dataset object,
+          if they have the same eval_dataset_var_name, but object ids are different,
+          then they will be assigned different name (via appending index to the
+          eval_dataset_var_name) when autologging.
+        * storing the dataset row samples, we need log them into metric_info artifact file.
+    (4) _metric_api_call_info, it is a double level map:
+       `metric_api_call_arg_dict_list_map[run_id]` wil get a map of
+        (metric_logged_key -> metric_call_command)
+        This data structure is used for:
+         * storing the call arguments dict for each metric call, we need log them into metric_info
+           artifact file.
+
+    Note: this class is not thread-safe.
+    Design rule for this class:
+     Because this class instance is a global instance, in order to prevent memory leak, it should
+     only holds IDs and other small objects references. This class internal data structure should
+     avoid reference to user dataset variables or model variables.
+    """
+    def __init__(self):
+        self._pred_result_id_to_dataset_name_and_run_id = {}
+        self._eval_dataset_info_map = defaultdict(lambda: defaultdict(list))
+        self._metric_api_call_info = defaultdict(dict)
+        self._log_post_training_metrics_enabled = True
+        self._metric_info_artifact_need_update = defaultdict(lambda: False)
+
+    def should_log_post_training_metrics(self):
+        """
+        Check whether we should run patching code for autologging post training metrics.
+        This checking should surround the whole patched code due to the safe guard checking,
+        See following note.
+
+        Note: It includes checking `_SklearnTrainingSession.is_active()`, This is a safe guarding
+        for cross validator/grid search case:
+          running cross validator/grid search, the nested model.fit will be called in parallel,
+          but, the _autolog_training_status is a global status without thread-safe lock protecting.
+          This safe guarding will prevent code run into this case.
+        """
+        return _SklearnTrainingSession.is_active() and self._log_post_training_metrics_enabled
+
+    def disable_log_post_training_metrics(self):
+        old_status = self._log_post_training_metrics_enabled
```

Comment:
I think we should compute `old_status` in the `LogPostTrainingMetricsDisabledScope.__enter__` method. Otherwise, what happens if the context manager isn't used for some time after being returned by `disable_log_post_training_metrics`?